### PR TITLE
Monitor miner failures to avoid failure spam

### PR DESCRIPTION
**Pull request by elijaharita**
I have a theory that some memory spikes could be caused by sudden bursts of failed requests to miners. This introduces a struct that tracks miner failures and temporarily suspends new retrievals from being started with miners that have generated too many errors in a short amount of time.

This'll also help reduce log spam a bit.

Currently the config struct is hardcoded at maximum 5 failures allowed within 15 seconds, with suspension duration of 1 minute. YAML options can be added in a subsequent PR.

I think @kylehuntsman might want to look into adding a metric for how many miners are currently suspended later (and maybe how many are properly working? I can share some info about how to access that info)

### CODE REVIEW

File: filecoin/minermonitor.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	failureHistoryDuration   time.Duration
+	suspensionDuration       time.Duration
+}
+
+// Miner monitor is a helper which assists in tracking which miners are not
+// responding and decides whether they should be temporarily suspended
+type minerMonitor struct {
+	cfg minerMonitorConfig
+
+	lk       sync.Mutex
```
Comment (hannahhoward) [inline filecoin/minermonitor.go:34]: not critical but this is a good usecase for a sync.RWLock

File: filecoin/minermonitor.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+func newMinerMonitor(cfg minerMonitorConfig) *minerMonitor {
+	return &minerMonitor{
+		cfg: cfg,
+
+		statuses: make(map[peer.ID]minerStatus),
+	}
+}
+
+func (monitor *minerMonitor) suspended(miner peer.ID) bool {
+	monitor.lk.Lock()
```
Comment (hannahhoward) [inline filecoin/minermonitor.go:47]: with this block needing only a read lock

File: filecoin/minermonitor.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		return status.suspended()
+	}
+
+	return false
+}
+
+func (monitor *minerMonitor) recordFailure(miner peer.ID) {
+	log.Warnf("%#v", monitor.cfg)
+
+	monitor.lk.Lock()
```
Comment (hannahhoward) [inline filecoin/minermonitor.go:60]: and this being staying a write lock as you have it now.

File: filecoin/minermonitor.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package filecoin
+
+import (
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type minerStatus struct {
```
Comment (hannahhoward) [inline filecoin/minermonitor.go:10]: non-blocking, but this is the sort of package you can can write a unit test for pretty quickly. Your logic looks really good (I think), but it's still the sort of thing that might have weird edge cases you're not catch.

**Review by hannahhoward — COMMENTED** [review]
All my comments are non-blocking. LGTM.

Just to confirm my understanding of your code logic, based on the configuration you have:
- if I get 5 failures in 15 seconds, I get suspended for 1 minute
- during the minute I'm suspended, there can potentially be a re-upp of my suspension if more failures come in (presumably this should drop off quickly cause code in the retriever stops going to the miner once its suspended, but maybe there are already things in flight)

**Review by hannahhoward — APPROVED** [review]

File: filecoin/minermonitor.go (hunk 4 above)
Comment (hannahhoward) [reply filecoin/minermonitor.go:10]: clearly, I remain the only programmer at PL who likes unit tests.

Comment (kylehuntsman) [reply filecoin/minermonitor.go:10]: > clearly, I remain the only programmer in PLN who likes unit tests.

Be the change you want to see in the world Hannah.

Comment (hannahhoward) [reply filecoin/minermonitor.go:10]: truthfully, and secretely, peerIDs are just strings -- 

here's code I use in another package for it:

```golang
var peerSeq int

// GeneratePeers creates n peer ids.
func GeneratePeers(n int) []peer.ID {
	peerIds := make([]peer.ID, 0, n)
	for i := 0; i < n; i++ {
		peerSeq++
		p := peer.ID(fmt.Sprint(peerSeq))
		peerIds = append(peerIds, p)
	}
	return peerIds
}
```

I wish packages that define these types would provide mocking generators. Sadly, this is not the case and I've encountered philosophical objections when I bring it up.
